api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -812,6 +821,14 @@ message Cluster {
   // of 0 would indicate that none of the timeout was used or that the timeout was infinite. A value
   // of 100 would indicate that the request took the entirety of the timeout given to it.
   bool track_timeout_budgets = 47;
+
+  // Optional customization and configuration of upstream type.
+  // By default, Envoy will create an upstream of type envoy.filters.upstreams.http.default
+  // The default upstream will create a envoy.filters.upstreams.http.tcp upstream
+  // if the request is a CONNECT request and connect termination is configured.
+  // Otherwise it will be envoy.filters.upstreams.http.http, with the protocol
+  // based on `http2_protocol_options`.
+  UpstreamType upstream_type = 48;
```

Comment:
The comment seems coupled with `GenericUpstream` defined in `router`, which is used only by "HttpConnManager".
Another [`GenericUpstream`](https://github.com/envoyproxy/envoy/blob/3adf14af5ca9f34221d59d693b173d20b27c1614/source/common/tcp_proxy/upstream.h#L44) is defined in TcpProxy. I assume it will be referred by `envoy.filters.upstreams.tcp.http` and `envoy.filters.upstreams.tcp.tcp`. 
This proto field `upstream_type` could be used by tcp_proxy, and also could be used by "http_conn_manager", and **both**
To address the cluster referred by both, the `upstream_type` is should be a collection type and at least 2 pairs 
Entry 1 key = `envoy.filters.upstreams.http`, value = `a value of UpstreamType`
Entry 2 key = `envoy.filters.upstreams.tcp`, value = `a value of UpstreamType`

Reply:
Hmm, rereading the comment I feel I had the wrong interpretation that `envoy.filters.upstreams.http.tcp` is referring to the class `TcpUpstream` in `extensions/upstreams/http/tcp/upstream_request.h`

Generally the new api looks good to me. I think I am able to rebase https://github.com/envoyproxy/envoy/pull/11112 on top of this API.